### PR TITLE
Slack with versioned context

### PR DESCRIPTION
**Pull request by liorblob**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://github.com/demisto/etc/issues/23544
fixes: https://github.com/demisto/etc/issues/25944

## Description
Added usage of versioned integration context.

Also some additional fixes:

- changed `format` to f-strings
- changed doc strings to google format
- added condition to not set the asyncio loop in long running execution

## Must have
- [ ] Tests WIP

### CODE REVIEW

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 1]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+        :type updated_list: ``list``
+        :param updated_list: The updated list.
+
+        :type key: ``str``
+        :param key: The key to replace elements by.
+
+        :rtype: ``list``
+        :return: The merged list.
+
+        Example:
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3839]: need to check how this looks in the script helper, i think we might need to move the examples to be above the args

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 2]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+        If the version is too old by the time the context is set, another attempt will be made until the limit.
+
+        :type context: ``dict``
+        :param context: A dictionary of keys and values to set.
+
+        :type object_keys: ``dict``
+        :param object_keys: A dictionary to map between context keys and their unique ID for merging them.
+
+        :type sync: ``bool``
+        :param sync: Whether to make a versioned request.
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3932]: need to update desc here

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 3]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+
+        :rtype: ``None``
+        :return: None
+        """
+        attempt = 0
+
+        # do while...
+        while True:
+            if attempt == CONTEXT_UPDATE_RETRY_TIMES:
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3941]: why isnt this an arg to the func ?

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 4]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+                raise Exception('Failed updating integration context. Max retry attempts exceeded.')
+
+            # Update the latest context and get the new version
+            integration_context, version = update_context(context, object_keys, sync)
+
+            demisto.info('Attempting to update the integration context with version {}.'.format(version))
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3947]:
```suggestion
            demisto.debug('Attempting to update the integration context with version {}.'.format(version))
```

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 5]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+            integration_context = latest_integration_context_versioned.get('context', {})
+            if sync:
+                version = latest_integration_context_versioned.get('version', 0)
+        else:
+            integration_context = latest_integration_context_versioned
+
+        return integration_context, version
+
+
+    def update_context(context, object_keys=None, sync=True):
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3987]:
```suggestion
    def update_integration_context(context, object_keys=None, sync=True):
```

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 6]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+                raise Exception('Failed updating integration context. Max retry attempts exceeded.')
+
+            # Update the latest context and get the new version
+            integration_context, version = update_context(context, object_keys, sync)
+
+            demisto.info('Attempting to update the integration context with version {}.'.format(version))
+
+            # Attempt to update integration context with a version.
+            # If we get a ValueError (DB Version), the version is too old. Then we need to try again.
+            attempt = attempt + 1
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3951]:
```suggestion
            attempt += 1
```

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 7]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+            if attempt == CONTEXT_UPDATE_RETRY_TIMES:
+                raise Exception('Failed updating integration context. Max retry attempts exceeded.')
+
+            # Update the latest context and get the new version
+            integration_context, version = update_context(context, object_keys, sync)
+
+            demisto.info('Attempting to update the integration context with version {}.'.format(version))
+
+            # Attempt to update integration context with a version.
+            # If we get a ValueError (DB Version), the version is too old. Then we need to try again.
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3950]:
```suggestion
            # If we get a ValueError (DB Version), then the version is not updated and we need to try again.
```

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 8]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+            attempt = attempt + 1
+            try:
+                set_integration_context(integration_context, sync, version)
+                demisto.info('Successfully updated integration context. New version is {}.'
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3954]:
```suggestion
                demisto.debug('Successfully updated integration context. New version is {}.'
```

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 9]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+                             .format(version + 1 if version != -1 else version))
+                break
+            except ValueError as ve:
+                demisto.info('Failed updating integration context with version {}: {} Attempts left - {}'
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3958]:
```suggestion
                demisto.debug('Failed updating integration context with version {}: {} Attempts left - {}'
```

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 10]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+                             .format(version, str(ve), CONTEXT_UPDATE_RETRY_TIMES - attempt))
+                # Sleep for a random time
+                time_to_sleep = randint(1, 100) / 1000
+                time.sleep(time_to_sleep)
+
+
+    def get_integration_context_with_version(sync):
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:3965]: default sync should be true?

File: Packs/Base/Scripts/CommonServerPython/CommonServerPython.py [hunk 11]
```diff
@@ -3815,5 +3816,207 @@ def dict_safe_get(dict_object, keys, default_return_value = None):
...
+
+        :rtype ``tuple``
+        :return: The updated integration context along with the current version.
+
+        """
+        integration_context, version = get_integration_context_with_version(sync)
+        if not object_keys:
+            object_keys = {}
+
+        for key, value in context.items():
```
Comment (Itay4) [inline Packs/Base/Scripts/CommonServerPython/CommonServerPython.py:4009]: `value` is not used

File: Tests/demistomock/demistomock.py [hunk 12]
```diff
@@ -5,6 +5,7 @@
 import uuid
 
 integrationContext = {}
+integrationContextVersioned = {}
```
Comment (Itay4) [inline Tests/demistomock/demistomock.py:8]: why do we need new var here? shouldnt we return the same object ?

File: Packs/Slack/Integrations/Slack/Slack.py [hunk 13]
```diff
@@ -1853,12 +1919,13 @@ def init_globals(command_name: str = ''):
     BOT_ICON_URL = demisto.params().get('bot_icon')  # Bot default icon url defined by the slack plugin (3-rd party)
     MAX_LIMIT_TIME = int(demisto.params().get('max_limit_time', '60'))
     PAGINATED_COUNT = int(demisto.params().get('paginated_count', '200'))
+    SYNC_CONTEXT = True
```
Comment (Itay4) [inline Packs/Slack/Integrations/Slack/Slack.py:1922]: why do we set it here and not just at the top?

File: Packs/Slack/Integrations/Slack/Slack.py [hunk 14]
```diff
@@ -186,8 +179,12 @@ def get_user_by_name(user_to_search: str, update_context: bool = True) -> dict:
 def search_slack_users(users) -> list:
```
Comment (Itay4) [inline Packs/Slack/Integrations/Slack/Slack.py:179]: add type hint for users

File: Packs/Slack/Integrations/Slack/Slack.py [hunk 15]
```diff
@@ -225,32 +224,12 @@ def find_mirror_by_investigation() -> dict:
...
-    for key, value in context.items():
-        demisto.debug(f'Slack - updating context value: {key} = {value}')
-        integration_context[key] = json.dumps(value)
-
-    demisto.info('Slack - Updating integration context.')
-    demisto.debug(f'Slack - integration context: {str(integration_context)}')
-    demisto.setIntegrationContext(integration_context)
-
-
 def set_name_and_icon(body, method):
```
Comment (Itay4) [inline Packs/Slack/Integrations/Slack/Slack.py:227]: add type hints

File: Packs/Slack/Integrations/Slack/Slack.py [hunk 16]
```diff
@@ -1044,16 +1063,19 @@ async def create_incidents(incidents: list, user_name: str, user_email: str, use
 async def listen(**payload):
```
Comment (Itay4) [inline Packs/Slack/Integrations/Slack/Slack.py:1063]: add type hints

File: Packs/Slack/Integrations/Slack/Slack.py [hunk 17]
```diff
@@ -1105,25 +1127,33 @@ async def listen(**payload):
...
                 investigation_id = mirror['investigation_id']
                 await handle_text(client, investigation_id, text, user)
         # Reset module health
         demisto.updateModuleHealth("")
     except Exception as e:
-        await handle_listen_error('Error occurred while listening to Slack: {}'.format(str(e)))
+        await handle_listen_error(f'Error occurred while listening to Slack: {str(e)}')
 
 
 async def get_user_by_id_async(client, user_id):
```
Comment (Itay4) [inline Packs/Slack/Integrations/Slack/Slack.py:1144]: add type hints

File: Packs/Slack/Integrations/Slack/Slack.py [hunk 18]
```diff
@@ -1912,6 +1979,10 @@ def main():
         if is_debug_mode():
             print_thread_dump()
 
+        if is_versioned_context_available():
+            # Refresh the integration context
+            demisto.getIntegrationContextVersioned(True)
```
Comment (Itay4) [inline Packs/Slack/Integrations/Slack/Slack.py:1984]: somewhat strange

File: Packs/Slack/Integrations/Slack/Slack.py [hunk 19]
```diff
@@ -6,15 +6,16 @@
 from slack.web.slack_response import SlackResponse
 
 from distutils.util import strtobool
+from random import randint
```
Comment (Itay4) [inline Packs/Slack/Integrations/Slack/Slack.py:9]: not used

File: Packs/Slack/Integrations/Slack/Slack.yml [hunk 20]
```diff
@@ -51,10 +51,10 @@ configuration:
 - defaultvalue: 'true'
   display: Long running instance. Required for investigation mirroring and direct
     messages.
+  hidden: true
```
Comment (Itay4) [inline Packs/Slack/Integrations/Slack/Slack.yml:54]: lets remove any mention of this param in the readme

File: Packs/Slack/Integrations/Slack/Slack.py (hunk 18 above)
Comment (liorblob) [reply Packs/Slack/Integrations/Slack/Slack.py:1984]: in progress

File: Packs/Slack/Integrations/Slack/Slack.yml (hunk 20 above)
Comment (liorblob) [reply Packs/Slack/Integrations/Slack/Slack.yml:54]: There is no mention of it in the readme

**Comment by Itay4** [comment]
@roysagi please review rn

**Comment by kirbles19** [comment]
@liorblob Done.

**Pull request merged**